opentracing/dapperish/main.go

```diff
@@ -0,0 +1,68 @@
+package main
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"html"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"os"
+	"runtime"
+
+	"github.com/opentracing/api-golang/opentracing"
+)
+
+func client() {
+	reader := bufio.NewReader(os.Stdin)
+	for {
+		span, ctx := opentracing.StartSpan("getInput")
+		span.Info("ctx: ", ctx)
+		fmt.Print("Enter text: ")
+		text, _ := reader.ReadString('\n')
+		span.Info(text)
+
+		httpClient := &http.Client{}
+		httpReq, _ := http.NewRequest("POST", "http://localhost:8080/", bytes.NewReader([]byte(text)))
+		opentracing.AddContextIDToHttpHeader(span.ContextID(), httpReq.Header)
+		resp, err := httpClient.Do(httpReq)
+		if err != nil {
+			span.Error("error: ", err)
+		} else {
+			span.Info("got response: ", resp)
+		}
+
+		span.Finish()
+	}
+}
+
+func server() {
+	http.HandleFunc("/", func(w http.ResponseWriter, req *http.Request) {
+		reqCtxID, err := opentracing.GetContextIDFromHttpHeader(
+			req.Header, opentracing.GlobalTracer())
+		if err != nil {
+			panic(err)
+		}
+
+		serverSpan, _ := opentracing.StartSpan("serverSpan", reqCtxID)
```

Comment:
> use an annotation on the client and/or server side to clarify that there's an RPC

I am ok with using annotations for representing structural concerns at the RAW layer, but given that this is a "construction" layer, I would prefer a strongly typed way of expressing structural relationships, otherwise every instrumenter needs to know the exact name of the annotation, and its exact value type, which is Zipkin's cs/cr mess all over again.

What probably makes the multi-hop case difficult is that it's essentially a data pipeline scenario, which, as we already said earlier, is indeed difficult in the span model. Again, I have the same issue with using custom tags to solve it, because construction layer should be explicit about structural relationships so that all instrumented frameworks produce the data consistently.

One way I think the span model can work is by recording explicit _elder sibling_ relationship. 

```
 client span
+------------------------------------------------+
    +------+   +------+   +-------------------+
     hop 1      hop 2      server
```
Each router records is "hop" span as the time spent in the router. Hop-2 must record not only that it's a child of "client span", but that it also strictly follows "hop-1", and "Server" strictly follows "hop-2". Then the clock skew adjustment can be done by assuming that all 4 network intervals between these spans are equal.  

Unfortunately, at the moment I do not see an easy way to describe that relationship, especially in the "server" span which has no knowledge that it was reached as a sibling of hop-2 and not as a direct child of "client span". Can I take you up on your offer to explain how you thought you could solve this with tags?